aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+
+Limiting client actions on the application needs to be done inside event handlers and not depend on the current rendered view state. 
+
+For example, imagine a counter where we want to only allow users to click the counter 3 times. If the logic for our counter is similar to this
+```
+<p>@count<p>
+@if(count < 3)
+{
+  <button @onclick="IncrementCount" value="Increment count" />
+}
+
+@code 
+{
+  int count = 0;
+  void IncrementCount() => count++;
+}
+```
+a client can dispatch one or more increment events before we produce a new render and result in the count being incremented over 3. The right way to achieve something like this is as follows.
+```
+<p>Count @count<p>
+@if(count < 3)
+{
+  <button @onclick="IncrementCount" value="Increment count" />
+}
+
+@code 
+{
+  int count = 0;
+  void IncrementCount() => count < 3 ? count++ : count;
+}
+```
+
+By adding the `count < 3` check inside the handler we force the decission to be taken based on the current app state, and not the state of the UI which might be out of date.
+
+# Other security recommendations
+In general, all the recommendations for securing ASP.NET Core applications apply to blazor server-side applications.
+
+In this section we will go through some of the common application threats, explain what blazor server-side does to address it and what recommended application specific actions can be taken to further protect your application.
+
+## Logging and sensitive data
+Blazor server-side applications interact heavily with the client through events and JS interop. All these interactions are recorded on the server logs via the use of `Microsoft.AspNetCore.Extensions.Logging.ILogger`. Blazor avoids logging sensitive information, like actual events or JS interop inputs/outputs.
+
+When an error occurs on the server we notify the client and tear down the session. By default the client receives a generic error message that can be read when opening the developer tools.
+
+This error does not include any callstack and doesn't provide any details of what the cause of the error was. The server logs contain such information and for development purposes it can be made available to the client by turning on detailed errors.
+
+This can be done either on `CircuitOptions.DetailedErrors` or by setting the `DetailedErrors` configuration key, for example through the `ASPNETCORE_DETAILEDERRORS=true` environment variable.
+
+## Protecting information in transit
+Blazor server-side uses SignalR under the hood to comunicate between the client and the server. That means that blazor server-side will normally use whatever transport SignalR negotiates (typically websockets).
+
+Blazor server-side doesn't take any extra step to ensure the integrity and confidentiality of the data sent between the server and the client.
+
+For that reason, the recommendation is to use HTTPS for communications if your app is not in a trusted environment.
+
+## Cross-site scripting (XSS)
```

Comment:
Should/could [trusted types](https://developers.google.com/web/updates/2019/02/trusted-types) be mentioned as a potential (future) mitigation for XSS? For more general notes [WICG/trusted-types](https://github.com/WICG/trusted-types/wiki/FAQ).